### PR TITLE
[fix](index compaction)support compact multi segments in one index

### PR DESCRIPTION
**Pull request by qidaye**

### CODE REVIEW

**Review by zzzxl1993 — APPROVED** [review]
LGTM

File: src/core/CLucene/index/IndexWriter.cpp [hunk 1]
```diff
@@ -1515,9 +1506,9 @@ void IndexWriter::addIndexesSegments(std::vector<lucene::store::Directory *> &di
         {
             SCOPED_LOCK_MUTEX(this->THIS_LOCK)
             for (auto dir: dirs) {
-                SegmentInfos sis;// read infos from dir
-                sis.read(dir);
-                segmentInfos->insert(&sis, true);
+                SegmentInfos* sis = new SegmentInfos(); // Create SegmentInfos on heap
```
Comment (xiaokang) [inline src/core/CLucene/index/IndexWriter.cpp:1509]: where is sis deleted?

File: src/core/CLucene/index/IndexWriter.cpp [hunk 2]
```diff
@@ -1246,39 +1246,30 @@ void IndexWriter::indexCompaction(std::vector<lucene::store::Directory *> &src_d
...
+    assert(readers.size() == numIndices);
 
     // check hasProx
     bool hasProx = false;
     {
         if (!readers.empty()) {
-            auto reader = static_cast<SegmentReader*>(readers[0]);
-            hasProx = reader->getFieldInfos()->hasProx();
+            IndexReader* reader = readers[0];
+            hasProx = reader->hasProx();
```
Comment (airborne12) [inline src/core/CLucene/index/IndexWriter.cpp:1270]: why from fieldinfo to reader level

File: src/core/CLucene/index/IndexWriter.cpp [hunk 3]
```diff
@@ -1246,39 +1246,30 @@ void IndexWriter::indexCompaction(std::vector<lucene::store::Directory *> &src_d
...
     //Set of IndexReaders
     if (infoStream != NULL) {
-        message(string("src index dir size: ") + Misc::toString(numSegments));
+        message(string("src index dir size: ") + Misc::toString(numIndices));
     }
-    for (int32_t i = 0; i < numSegments; i++) {
-        SegmentInfo *si = segmentInfos->info(i);
-        IndexReader *reader = SegmentReader::get(si, MERGE_READ_BUFFER_SIZE, false /* mergeDocStores */);
+    for (int32_t i = 0; i < numIndices; i++) {
+        IndexReader* reader = lucene::index::IndexReader::open(src_dirs[i], MERGE_READ_BUFFER_SIZE, false);
```
Comment (xiaokang) [inline src/core/CLucene/index/IndexWriter.cpp:1257]: add comment to explain why change si to src_dir[i]

File: src/core/CLucene/index/IndexWriter.cpp [hunk 4]
```diff
@@ -1618,11 +1606,25 @@ void IndexWriter::mergeTerms(bool hasProx) {
...
+        if (infoStream != nullptr) {
+            std::wstring ws = smallestTerm->text();
+            std::wstring f = smallestTerm->field();
+            std::string name = std::string(ws.begin(), ws.end());
+            std::string field = std::string(f.begin(), f.end());
+            message("smallestTerm name: " + name);
+            message("smallestTerm field: " + field);
+
+            if (top != nullptr) {
+                Term* topTerm = top->term;
```
Comment (xiaokang) [inline src/core/CLucene/index/IndexWriter.cpp:1619]: Are these wstring and string variables just for debug message? If true, add debug check for performance.

Comment (qidaye) [reply src/core/CLucene/index/IndexWriter.cpp:1619]: infoStream is null by default.
It's a common use in CLucene for debug. There is not a performance issue.

File: src/core/CLucene/index/IndexWriter.cpp (hunk 2 above)
Comment (qidaye) [reply src/core/CLucene/index/IndexWriter.cpp:1270]: Remove this API and use `reader->getFieldInfos()->hasProx()` instead.

File: src/core/CLucene/index/IndexWriter.cpp (hunk 3 above)
Comment (qidaye) [reply src/core/CLucene/index/IndexWriter.cpp:1257]: Done.

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by airborne12 — APPROVED** [review]
LGTM

**Comment by zzzxl1993** [comment]
LGTM

**Pull request merged**